### PR TITLE
Refactor Weights to remove awareness of types

### PR DESCRIPTION
**Pull request by decentralion**
I'm currently on a quest to separate cred computation away from any
plugin metadata (see #1557). This means we need a way to represent node
and edge weights without any explicit concept of 'types'.

This commit is a first step towards that. It removes the distinction
between 'type weights' and 'manual weights' in the weights data type.
Instead, we now just have node weights and edge weights. In contrast to
before, all weights are now interpreted as prefix matchers, e.g. a
single node or edge may match multiple weights; when this occurs, the
weights compose multiplicatively.

Since types were already identified by prefix, if a plugin wants to
assign a weight to a particular type, it may do so by specifying a
weight for that type's prefix. As before, it's possible to have a
type-level weight and a weight on a specific node, and compose them
multiplicatively.

As an added bonus, we could now sensibly have 'plugin-level' weights and
'type-level weights' and compose them multiplicatively. Thus, if we
realized that the Foo plugin is undervalued relative to the Bar plugin,
we could increase the Foo weight rather than needing to adjust all of
its types individually.

So as to keep the scope for this commit somewhat manageable, I modified
the underlying data type for Weights, but not any of the cred
computation interfaces. The weights pipeline still takes the plugin
declarations, and we still get the default type level weights from the
plugin's types. A future commit will modify the pipeline so that the
plugins provide default types alongside the Graph.

I deliberately did not provide an upgrade handler for the old style
weights JSON. This is sensible as the semantics are now different. In
the past, it was possible to specify a weight for a single node without
affecting the weights of other nodes whose addresses have the first
node's address as a prefix. Since this is no longer possible, there is
no universally "correct" way to handle the old weights files. In
practice, there are so few users that it is not a big deal either way.

Test plan:
This change has implications across the codebase and UI. In addition to
`yarn test --full` passing, I verified that:
- updating and recomputing works in the mainline UI
- updating and recomputing works in the legacy UI
- downloading weights from the UI and then explicitly loading them still
works

### CODE REVIEW

**Comment by decentralion** [comment]
cc @vsoch since this is the first step of the algorithm cleanup work that I mentioned. this change is a little messy because it propagates out to the UI code, but i encourage you to poke around a bit in the weights module and the weightEvaluator module--both will be relevant for understanding how cred calculation works.

**Comment by Beanow** [comment]
> In the past, it was possible to specify a weight for a single node without
> affecting the weights of other nodes whose addresses have the first
> node's address as a prefix. Since this is no longer possible, there is
> no universally "correct" way to handle the old weights files.

I think this is interesting and surfaces a somewhat unintuitive constraint on Node / Edge addresses. They shouldn't be prefixes of each other. Right now our reference plugins have this as a convention. For a future with independently developed plugins this may be tricky.

For example, we use a type prefix which prevents trouble here: 
```
https://github.com/sourcecred-test/example-github/issues/2
NodeAddress[sourcecred,github,ISSUE,sourcecred-test,example-github,2]

(Not a prefix of)

https://github.com/sourcecred-test/example-github/issues/2#issuecomment-373768703
NodeAddress[sourcecred,github,COMMENT,ISSUE,sourcecred-test,example-github,2,373768703]
```

But that wouldn't work if it was:
```
https://github.com/sourcecred-test/example-github/issues/2
NodeAddress[sourcecred,github,ISSUE,sourcecred-test,example-github,2]

(Is a prefix of)

https://github.com/sourcecred-test/example-github/issues/2#issuecomment-373768703
NodeAddress[sourcecred,github,ISSUE,sourcecred-test,example-github,2,COMMENT,373768703]
```

Initially I was going to suggest "perhaps prefix matching should be optional", but even when you opt-in for prefix matching this could easily catch someone off-guard. So perhaps we should have a runtime validation of the graph when computing cred, which warns about addresses that are prefixes of each other?

File: src/analysis/weights.js [hunk 1]
```diff
@@ -20,19 +20,15 @@ export type NodeWeight = number;
 export type EdgeWeight = {|+forwards: number, +backwards: number|};
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/analysis/weights.js]: Fixed.

File: src/analysis/weightEvaluator.js [hunk 2]
```diff
@@ -1,65 +1,84 @@
...
 
-import * as NullUtil from "../util/null";
 import type {NodeAddressT, EdgeAddressT} from "../core/graph";
 import type {NodeType, EdgeType} from "./types";
-import type {Weights, EdgeWeight} from "./weights";
+import type {Weights as WeightsT, EdgeWeight} from "./weights";
+import * as Weights from "./weights";
 import {NodeTrie, EdgeTrie} from "../core/trie";
 
 export type NodeWeightEvaluator = (NodeAddressT) => number;
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/analysis/weightEvaluator.js]: Done.

File: src/analysis/weightEvaluator.js [hunk 3]
```diff
@@ -1,65 +1,84 @@
...
-  const nodeTrie = new NodeTrie();
+  const {nodeWeights} = Weights.copy(weights);
+
   for (const {prefix, defaultWeight} of types) {
-    const weight = NullUtil.orElse(typeWeights.get(prefix), defaultWeight);
+    if (!nodeWeights.has(prefix)) {
+      nodeWeights.set(prefix, defaultWeight);
+    }
+  }
+  const nodeTrie = new NodeTrie();
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/analysis/weightEvaluator.js]: Done.

File: src/analysis/weightEvaluator.js [hunk 4]
```diff
@@ -1,65 +1,84 @@
...
+  for (const [prefix, weight] of nodeWeights.entries()) {
     nodeTrie.add(prefix, weight);
   }
   return function nodeWeight(a: NodeAddressT): number {
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/analysis/weightEvaluator.js]: Done.

File: src/analysis/weightEvaluator.js [hunk 5]
```diff
@@ -1,65 +1,84 @@
...
-  const typeWeights = weights.edgeTypeWeights;
-  const edgeTrie = new EdgeTrie();
+  const {edgeWeights} = Weights.copy(weights);
   for (const {prefix, defaultWeight} of types) {
-    const weight = NullUtil.orElse(typeWeights.get(prefix), defaultWeight);
+    if (!edgeWeights.has(prefix)) {
+      edgeWeights.set(prefix, defaultWeight);
+    }
+  }
+  const edgeTrie = new EdgeTrie();
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/analysis/weightEvaluator.js]: Done

File: src/analysis/weightEvaluator.js [hunk 6]
```diff
@@ -1,65 +1,84 @@
...
+  for (const [prefix, weight] of edgeWeights.entries()) {
     edgeTrie.add(prefix, weight);
   }
   return function evaluator(address: EdgeAddressT) {
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/analysis/weightEvaluator.js]: Done.

**Pull request merged**